### PR TITLE
feat(release): Add epoch start estimator script

### PR DESCRIPTION
**Pull request by VanBarbascu**
Usage:
```
python3 estimate_epoch_start_time.py --chain_id mainnet --num_future_epochs 10 --num_past_epochs 20
```

<details><summary>Example output</summary>

```
Epoch -1: 14 hours, 59 minutes                                                                                                                                                                                        
Epoch -2: 14 hours, 58 minutes                                                                                                                                                                                        
Epoch -3: 13 hours, 22 minutes                                                                                                                                                                                        
Epoch -4: 13 hours, 22 minutes                                                                                                                                                                                        
Epoch -5: 13 hours, 42 minutes                                                                                                                                                                                        
Epoch -6: 13 hours, 27 minutes                                                                                                                                                                                        
Epoch -7: 13 hours, 53 minutes                                                                                                                                                                                        
Epoch -8: 13 hours, 22 minutes
Epoch -9: 13 hours, 46 minutes
Epoch -10: 13 hours, 30 minutes
Epoch -11: 13 hours, 20 minutes
Epoch -12: 14 hours, 11 minutes
Epoch -13: 13 hours, 58 minutes
Epoch -14: 13 hours, 58 minutes
Epoch -15: 13 hours, 19 minutes
Epoch -16: 13 hours, 49 minutes
Epoch -17: 13 hours, 32 minutes
Epoch -18: 13 hours, 16 minutes
Epoch -19: 13 hours, 18 minutes
Epoch -20: 13 hours, 18 minutes

Exponential weighted average epoch length: 13 hours, 52 minutes
Predicted start of epoch 1: 2024-09-25 18:18:07 Wednesday
Predicted start of epoch 2: 2024-09-26 08:10:37 Thursday
Predicted start of epoch 3: 2024-09-26 22:03:07 Thursday
Predicted start of epoch 4: 2024-09-27 11:55:37 Friday
Predicted start of epoch 5: 2024-09-28 01:48:07 Saturday
Predicted start of epoch 6: 2024-09-28 15:40:37 Saturday
Predicted start of epoch 7: 2024-09-29 05:33:07 Sunday
Predicted start of epoch 8: 2024-09-29 19:25:37 Sunday
Predicted start of epoch 9: 2024-09-30 09:18:07 Monday
Predicted start of epoch 10: 2024-09-30 23:10:37 Monday
``` 

</details>

### CODE REVIEW

File: debug_scripts/estimate_epoch_start_time.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+                                    time.gmtime(future_timestamp))
+        future_epochs.append(future_date)
+
+        print(f"Predicted start of epoch {i}: {future_date}")
+
+    return future_epochs
+
+
+# First, we get the latest block and save the next_epoch_id
+def get_latest_block(url):
```
Comment (tayfunelmas) [inline debug_scripts/estimate_epoch_start_time.py:92]: nit: move this closer to get_block_by_hash definition.

File: debug_scripts/estimate_epoch_start_time.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+                          exponential_weighted_average_epoch_length,
+                          args.num_future_epochs)
+
+
+# Set up command-line argument parsing
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(
+        description="Approximate future epoch start dates for NEAR Protocol.")
+    parser.add_argument("--url", required=True, help="The RPC URL to query.")
+    parser.add_argument("--num_epochs",
```
Comment (tayfunelmas) [inline debug_scripts/estimate_epoch_start_time.py:128]: I would rename this to num_past_epochs to be more clear

File: debug_scripts/estimate_epoch_start_time.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    }
+    response = requests.post(url, json=payload).json()
+    return response['result']['header']
+
+
+# Main function to run the process
+def main(args):
+    latest_block = get_latest_block(args.url)
+    next_epoch_id = latest_block['next_epoch_id']
+    current_epoch_first_block = get_block_by_hash(args.url, next_epoch_id)
```
Comment (tayfunelmas) [inline debug_scripts/estimate_epoch_start_time.py:109]: does not this block belong to the previous epoch?

File: debug_scripts/estimate_epoch_start_time.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    )
+
+    return epoch_lengths, exponential_weighted_average_epoch_length
+
+
+# Function to approximate future epoch start dates
+def predict_future_epochs(starting_epoch_timestamp, avg_epoch_length,
+                          num_future_epochs):
+    future_epochs = []
+    current_timestamp = starting_epoch_timestamp / 1e9  # Convert from nanoseconds to seconds
```
Comment (tayfunelmas) [inline debug_scripts/estimate_epoch_start_time.py:75]: nit: maybe for readability you can define a small function to convert epoch timestamp to seconds.

**Review by tayfunelmas — APPROVED** [review]

**Comment by tayfunelmas** [comment]
Not needed for this PR but it would be nice to have an argument to set a specific timezone.

File: debug_scripts/estimate_epoch_start_time.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+
+    print(
+        f"\nExponential weighted average epoch length: {exponential_weighted_average_epoch_length} seconds"
+    )
+
+    return epoch_lengths, exponential_weighted_average_epoch_length
+
+
+# Function to approximate future epoch start dates
+def predict_future_epochs(starting_epoch_timestamp, avg_epoch_length,
```
Comment (jancionear) [inline debug_scripts/estimate_epoch_start_time.py:72]: One observation - the tool predicts that the next epoch will start in 11 hours, but debug-ui says it'll happen in 13 hours.
I wonder where the discrepancy comes from :thinking: Maybe debug-ui doesn't use exponential average. /cc @robin-near

File: debug_scripts/estimate_epoch_start_time.py [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+                        type=float,
+                        default=0.1,
+                        help="Decay rate for exponential weighting.")
+    parser.add_argument("--num_future_epochs",
+                        type=int,
+                        default=3,
+                        help="Number of future epochs to predict.")
+
+    args = parser.parse_args()
+    main(args)
```
Comment (jancionear) [inline debug_scripts/estimate_epoch_start_time.py:142]: Very useful, thank you for making this!

To resolve https://github.com/near/nearcore/issues/12107 I would love to have two more things:
* If the voting date is X, when will the protocol upgrade happen?
* Given a day of the voting date find the hour that maximizes the probability that the upgrade happens during working hours

But that can be done in follow up PRs

File: debug_scripts/estimate_epoch_start_time.py (hunk 3 above)
Comment (jancionear) [reply debug_scripts/estimate_epoch_start_time.py:109]: +1 I think this is `prev_epoch_last_block`, but for this tool it doesn't really matter.

https://nomicon.io/BlockchainLayer/EpochManager/Epoch#epoch-id

> EpochId is defined as [...] for epoch with height T+2 it's the hash of the last block in epoch T

File: debug_scripts/estimate_epoch_start_time.py [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+        # Fetch the block data for start of previous epoch
+        next_block_data = get_block_by_hash(url, next_epoch_hash)
+        next_timestamp = int(next_block_data['timestamp'])
+
+        # Calculate the length of the epoch in nanoseconds
+        epoch_length = current_timestamp - next_timestamp
+        epoch_length_seconds = epoch_length / 1e9  # Convert to seconds
+        epoch_lengths.append(epoch_length_seconds)
+
+        print(f"Epoch {i+1}: {epoch_length_seconds} seconds")
```
[the remark this answers is not in the document]
Comment (VanBarbascu) [reply debug_scripts/estimate_epoch_start_time.py:48]: This is not straight forward.

File: debug_scripts/estimate_epoch_start_time.py (hunk 3 above)
Comment (VanBarbascu) [reply debug_scripts/estimate_epoch_start_time.py]: I renamed the variables.

**Pull request merged**